java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java

```diff
@@ -319,4 +321,79 @@ static void setBitMaskedByte(ArrowBuf data, int byteIndex, byte bitMask) {
     currentByte |= bitMask;
     data.setByte(byteIndex, currentByte);
   }
+
+  /**
+   * Concat two validity buffers.
+   * @param input1 the first validity buffer.
+   * @param numBits1 the number of bits in the first validity buffer.
+   * @param input2 the second validity buffer.
+   * @param numBits2 the number of bits in the second validity buffer.
+   * @param output the ouput validity buffer. It can be the same one as the first input.
+   */
+  public static void concatBits(ArrowBuf input1, int numBits1, ArrowBuf input2, int numBits2, ArrowBuf output) {
+    int numBytes1 = DataSizeRoundingUtil.divideBy8Ceil(numBits1);
+    int numBytes2 = DataSizeRoundingUtil.divideBy8Ceil(numBits2);
+    int numBytesOut = DataSizeRoundingUtil.divideBy8Ceil(numBits1 + numBits2);
+
+    if (BoundsChecking.BOUNDS_CHECKING_ENABLED) {
+      output.checkBytes(0, numBytesOut);
+    }
+
+    // copy the first bit set
+    if (input1 != output) {
+      PlatformDependent.copyMemory(input1.memoryAddress(), output.memoryAddress(), numBytes1);
+    }
+
+    if (bitIndex(numBits1) == 0) {
+      // The number of bits for the first bit set is a multiple of 8, so the boundary is at byte boundary.
+      // For this case, we have a shortcut to copy all bytes from the second set after the byte boundary.
+      PlatformDependent.copyMemory(input2.memoryAddress(), output.memoryAddress() + numBytes1, numBytes2);
+      return;
+    }
+
+    // the number of bits to fill a full byte after the first input is processed
+    int numBitsToFill = 8 - bitIndex(numBits1);
+
+    // mask to clear high bits
+    int mask = (1 << (8 - numBitsToFill)) - 1;
+
+    int numFullBytes = numBits2 / 8;
+
+    int prevByte = output.getByte(numBytes1 - 1) & mask;
+    for (int i = 0; i < numFullBytes; i++) {
+      int curByte = input2.getByte(i) & 0xff;
+
+      // first fill the bits to a full byte
+      int byteToFill = (curByte << (8 - numBitsToFill)) & 0xff;
+      output.setByte(numBytes1 + i - 1, byteToFill | prevByte);
+
+      // fill remaining bits in the current byte
+      // note that it is also the previous byte for the next iteration
+      prevByte = curByte >>> numBitsToFill;
+    }
+
+    // clear high bits for the previous byte, as it may be the last byte
+    int lastOutputByte = prevByte;
+
+    // the number of extra bits for the second input, relative to full bytes
+    int numRemainingBits = bitIndex(numBits2);
+
+    // process remaining bits from input2
+    if (numRemainingBits > 0) {
+      int remByte = input2.getByte(numBytes2 - 1) & 0xff;
+
+      int byteToFill = remByte << (8 - numBitsToFill);
+      lastOutputByte |= byteToFill;
+
+      if (numRemainingBits > numBitsToFill) {
```

Comment:
i think the logic would be clearer if you moved this if statement. as the last piece of logic after output.setByte on line 397.  Otherwise it seems strange be setting the non-ending byte as the last byte.

Reply:
I agree with you that it is weird that last two byte are written in reverse order. 

I have revised the code to solve this problem. Please take a look. Thanks a lot.